### ISSUE TITLE
`attr_readonly`: surprising inconsistency between what's saved and what's in memory

### ISSUE THREAD

**Issue by ghiculescu** [rails/rails#41520]
### Steps to reproduce

When an attribute is marked as `attr_readonly`, updates to it are not persisted. But, surprisingly, the in memory object *is* updated. There is no warning about this inconsistency. This means that if you do something that refers to this attribute's value after the update attempt, you'll be working with an incorrect value.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.text :content
    t.text :title
  end
end

class Post < ActiveRecord::Base
  attr_readonly :content
end

class BugTest < Minitest::Test
  def test_readonly
    post = Post.create!(content: "v1", title: "title v1")
    assert_equal "v1", post.content
    assert_equal "title v1", post.title
    assert_equal "v1", Post.find(post.id).content
    assert_equal "title v1", Post.find(post.id).title

    post = Post.find(post.id) # reload the object
    post.update(content: "v2", title: "title v2")
    assert post.valid?
    assert_equal "v2", post.content # passes
    assert_equal "title v2", post.title # passes
    assert_equal "title v2", Post.find(post.id).title # passes
    assert_equal "v2", Post.find(post.id).content # fails; it's still "v1"
  end

  def test_readonly_via_assign
    post = Post.create!(content: "v1", title: "title v1")
    post.content = "v2"
    post.title = "title v2"
    assert post.valid?
    post.save!

    assert_equal "v2", post.content # passes
    assert_equal "title v2", post.title # passes
    assert_equal "title v2", Post.find(post.id).title # passes
    assert_equal "v2", Post.find(post.id).content # fails; it's still "v1"
  end
end
```

### Expected behavior
I would expect the in memory object (`post.content`) to match the object in the database (`Post.find(post.id).content`). Or at least I'd expect an error or warning.

I am happy to put together a PR to improve this, I'd like feedback on which direction to go in. My ideas, in order of preference:

- Raise an error when assigning a readonly attribute. (Can be configured between logging or raising an error, similar to `action_on_strict_loading_violation` etc)
- Make the in-memory update also ignore readonly attributes, to match the persistence behavior.
- Make this more explicit in [the docs](https://api.rubyonrails.org/classes/ActiveRecord/ReadonlyAttributes/ClassMethods.html).
- Open to other suggestions!

### System configuration
**Rails version**: main branch

**Ruby version**: 2.7.2

**Comment by ghiculescu** [comment]
> what would `post.update(content: "v2", title: "hi")` now become? does it return true but only updated the title? that can be very misleading.

That's currently what happens. I updated the PR body with a test case that shows this.

>  i would expect it to be true, then have a validation error which says "content is readonly".

I'd be okay with a validation error too. Added a test case to confirm there isn't one currently.

**Comment by krtschmr** [comment]
validation_error would be okay, with `save!` or `update!` we would then get an exception.

i personally do like the exception once you even try to write into. if that's configurable, we have all cases covered in one

**Comment by ghiculescu** [comment]
I made a discuss thread about this https://discuss.rubyonrails.org/t/attr-readonly-confusing-behavior/77258 but unless anyone thinks it's a terrible idea, I plan to work on this soon.

**Comment by krtschmr** [comment]
nobody wants to discuss. how to proceed? @rafaelfranca do you have an opinion?

**Comment by krtschmr** [comment]
Should not be stale (:

**Comment by zzak** [comment]
@ghiculescu Still interested in working on this?

**Comment by ghiculescu** [comment]
@zzak I am, but it's not a high priority. Won't be sad if you go for it.

**Comment by zzak** [comment]
No rush! I just didn't want it to get auto-closed if you were still planning to pick it up. 🙏

**Comment by zzak** [comment]
@ghiculescu Is this issue not closed because #42705 hasn't been merged yet? :pray:

**Comment by ghiculescu** [comment]
Yeah that's right.
